src/libraries/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```diff
@@ -56,32 +57,120 @@ public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mo
                     eventFlags,
                     (uint)EventWaitHandleRights.FullControl);
 
-                ValidateHandle(handle, name, out createdNew);
+                int errorCode = Marshal.GetLastWin32Error();
 
-                EventWaitHandle ewh = new EventWaitHandle(initialState, mode);
-                SafeWaitHandle old = ewh.SafeWaitHandle;
-                ewh.SafeWaitHandle = handle;
-                old.Dispose();
+                if (handle.IsInvalid)
+                {
+                    handle.SetHandleAsInvalid();
+
+                    if (!string.IsNullOrEmpty(name) && errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    {
+                        throw new WaitHandleCannotBeOpenedException(SR.Format(SR.WaitHandleCannotBeOpenedException_InvalidHandle, name));
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+                }
+
+                createdNew = (errorCode != Interop.Errors.ERROR_ALREADY_EXISTS);
 
-                return ewh;
+                return CreateAndReplaceHandle(handle);
             }
         }
 
-        private static void ValidateHandle(SafeWaitHandle handle, string? name, out bool createdNew)
+        /// <summary>
+        /// Opens a specified named event wait handle, if it already exists, applying the desired access rights.
+        /// </summary>
+        /// <param name="name">The name of the event wait handle to be opened. If it's prefixed by "Global", it refers to a machine-wide event wait handle. If it's prefixed by "Local", or doesn't have a prefix, it refers to a session-wide event wait handle. Both prefix and name are case-sensitive.</param>
+        /// <param name="rights">The desired access rights to apply to the returned event wait handle.</param>
+        /// <returns>An existing named event wait handle.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="name"/> is an empty string.</exception>
+        /// <exception cref="WaitHandleCannotBeOpenedException">The named event wait handle does not exist or is invalid.</exception>
+        /// <exception cref="IOException">The path was not found.
+        /// -or-
+        /// A Win32 error occurred.</exception>
+        /// <exception cref="UnauthorizedAccessException">The named event wait handle exists, but the user does not have the security access required to use it.</exception>
+        public static EventWaitHandle OpenExisting(string name, EventWaitHandleRights rights)
```

Comment:
This method and others will need a supported attribute applied, as they only work on Windows, correct?

Reply:
Yes looks like added assembly level https://github.com/dotnet/runtime/blob/580e9ddc381b3521fea31d2c7746c18ef2c98629/src/libraries/System.Threading.AccessControl/Directory.Build.props#L5

Reply:
That was my assumption since these APIs were already Windows specific. But thanks for the reminder Stephen, and for confirming, Buyaa.